src/range_proof.rs

```diff
@@ -1,19 +1,18 @@
-use bulletproofs::r1cs::{ConstraintSystem, R1CSError};
+use bulletproofs::r1cs::{ConstraintSystem, LinearCombination, R1CSError};
 use curve25519_dalek::scalar::Scalar;
-use value::AllocatedQuantity;
 
 /// Enforces that the quantity of v is in the range [0, 2^n).
 pub fn range_proof<CS: ConstraintSystem>(
     cs: &mut CS,
-    v: AllocatedQuantity,
+    mut v: LinearCombination,
+    v_assignment: Option<u64>,
```

Comment:
would be cool if our `LinearCombination` types had their own assignment slots

Reply:
i suspect we'd have to add these into ZkVM `Expression` types, which is another reason to push them into BP's LC.

Reply:
Could you deduce the assignment of the `LinearCombination` by reaching into the constraint system and getting the variable assignment for each of its variables? (It's not very ergonomic, but I am more wondering if this is a proposal for a "new capability" or a proposal for a "nice wrapper for an existing capability"?)

Reply:
i can, but it'll cross through crates, so we'd need a public API to ask CS to give an assignment for an LC. IMHO it's cleaner to just keep the assignment there. It'll be the same cost for prover (they have to compute these anyway) and same zero cost for verifier.

Reply:
I mean, instead of providing a public API to have a roundabout way to get an assignment for a given LC, we probably should just keep the assignment within the LC itself.

Also: in the immediate protocol on top of BP we create a wrapper around `r1cs::Variable` to keep the assignment close to the variable (`AllocatedQuantity`), which is another hint that we probably should keep those together for both individual variables and for LCs.

Reply:
that makes sense, thanks for explaining!